Hosting Dashboard: Add PurchaseNotice to purchase settings page

sirbrillig opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Fixes https://linear.app/a8c/issue/SHILL-1081/add-purchasenotice-to-hosting-dashboard-purchase-settings

## Proposed Changes

This PR adds a notice component to the top of the purchase settings page which covers a number of scenarios. This just copies [the existing behavior from the calypso purchase settings page](https://github.com/Automattic/wp-calypso/blob/4d8f3fa385e796d0b57de640c67e34554bc3585e/client/me/purchases/manage-purchase/notices.tsx#L87).

Some examples:

<img width="711" height="430" alt="Screenshot 2025-09-02 at 4 07 42 PM" src="https://github.com/user-attachments/assets/2197ab6b-c9a2-420b-b8ed-0c898a07206a" />

<img width="745" height="465" alt="Screenshot 2025-09-02 at 4 29 37 PM" src="https://github.com/user-attachments/assets/116d5a20-7214-4d60-9ac0-472ca4f32750" />

<img width="790" height="450" alt="Screenshot 2025-09-01 at 4 35 09 PM" src="https://github.com/user-attachments/assets/f40548fb-7f5e-4443-a1a6-ddd33271f158" />

<img width="730" height="329" alt="Screenshot 2025-09-02 at 1 02 55 PM" src="https://github.com/user-attachments/assets/7051d453-3eac-4654-b411-201775bb6e3d" />

<img width="720" height="307" alt="Screenshot 2025-09-02 at 1 02 43 PM" src="https://github.com/user-attachments/assets/f9fe01d7-9a63-42b9-923d-aa7d14b0a657" />

<img width="708" height="341" alt="Screenshot 2025-09-02 at 1 02 32 PM" src="https://github.com/user-attachments/assets/fb8773d7-2c79-4ce7-b346-eb422d316f1b" />

<img width="698" height="299" alt="Screenshot 2025-09-02 at 2 55 03 PM" src="https://github.com/user-attachments/assets/063ada48-eed9-4fd2-9c8c-691128d2e237" />

<img width="691" height="293" alt="Screenshot 2025-09-26 at 1 02 08 PM" src="https://github.com/user-attachments/assets/1d563b14-a04a-4794-b86d-f58bedfe96be" />

<img width="699" height="298" alt="Screenshot 2025-09-26 at 1 02 33 PM" src="https://github.com/user-attachments/assets/ea8c12a1-828a-4484-ba0e-0b806e11c930" />

<img width="901" height="549" alt="Screenshot 2025-09-26 at 1 02 41 PM" src="https://github.com/user-attachments/assets/5863e2b1-fb00-468f-9439-fa109670df3e" />

## Testing Instructions

This is a little difficult to test as it requires having subscriptions that are in various states. Some scenarios to test this if you can arrange them:

- A used concierge session.
- An expired subscription (that's still active or it won't show in the list).
- An in-app purchase.
- An E-commerce plan trial.
- A subscription with a pending async payment.
- A subscription with an attached credit card that is expiring.
- A subscription without a payment method or in [one of the 8 similar notifications](https://github.com/Automattic/wp-calypso/blob/4d8f3fa385e796d0b57de640c67e34554bc3585e/client/me/purchases/manage-purchase/notices.tsx#L412-L836) some of which include a link to a dialog that shows other expiring subscriptions.

sirbrillig committed:
Add initial PurchaseWarnings

sirbrillig committed:
Add IAP purchase notice

sirbrillig committed:
Add more things in PurchaseNotice

sirbrillig committed:
Add PurchaseExpiringNotice

sirbrillig committed:
Move PurchaseNotice down to main body

sirbrillig committed:
Use snackbar for reinstallPlugins success/error

sirbrillig committed:
Move PurchaseExpiringNotice to own file

sirbrillig committed:
Add CreditCardExpiringNotice

sirbrillig committed:
Fix button styles for RenewNoticeAction

sirbrillig committed:
Add OtherRenewablePurchasesNotice

sirbrillig committed:
Simplify shouldShowOtherRenewablePurchasesNotice

sirbrillig committed:
Add renewableSitePurchases

sirbrillig committed:
Add UpcomingRenewalsDialog

sirbrillig committed:
Fix "Renew now" button style in RenewNoticeAction

sirbrillig committed:
Remove noticeActionOnClick for "Update all"; they already have a URL

fushar commented on client/dashboard/me/billing-purchases/purchase-settings/renew-notice-action.tsx:
```diff
@@ -0,0 +1,80 @@
+			</Button>
+		);
+	}
+	return null;
```
I think this case should never happen -- `shouldShowRenewNoticeAction()` above should cover all cases. However it seems to still happen based on the screenshots in this PR description. It causes extra margin/padding at the bottom, because the notice is trying to render an action button but we return null:

<img width="675" height="89" alt="Image" src="https://github.com/user-attachments/assets/075d8997-10e7-4f58-a980-0b1fbe5a609b" />

fushar commented on client/dashboard/me/billing-purchases/purchase-settings/index.tsx:
```diff
@@ -464,19 +466,20 @@ function ReinstallButton( { purchase }: { purchase: Purchase } ) {
-							reinstallPlugins();
+							reinstallPlugins( undefined, {
+								onSuccess: () => {
+									createSuccessNotice( __( 'Plugins reinstalled.' ), { type: 'snackbar' } );
```
Let's use the new `meta` mechanism:

- https://github.com/Automattic/wp-calypso/pull/105880

sirbrillig replied:
I'm not sure how to use the error handler with an error passed back from the server. 🤔 I've asked about it in Slack here: p1759176105963599-slack-C0982082MSR 

https://github.com/Automattic/wp-calypso/blob/d30b8f738deda5ec3f7dd787e9235a2ba9c394a5/client/dashboard/me/billing-purchases/purchase-settings/index.tsx#L473-L477

But I suppose I could just use the generic message. I'll just do that for now.

fushar reviewed (commented):
I'm not really comfortable reviewing all the logic here; I think it's better for Shilling team 😄 left some comments; I think in general I don't spot anything extraordinary in the code.

arthur791004 commented on client/dashboard/me/billing-purchases/purchase-settings/upcoming-renewals-dialog.tsx:
```diff
@@ -0,0 +1,141 @@
+					<VStack key={ purchase.ID }>
+						<HStack>
+							<HStack alignment="left">
+								<CheckboxControl
```
Should we use `DataForm` here with custom Edit component to render row for consistency 🤔

sirbrillig replied:
Good idea. In fact, I think I'm going to use a DataViews.

arthur791004 commented on client/dashboard/me/billing-purchases/purchase-settings/upcoming-renewals-dialog.tsx:
```diff
@@ -0,0 +1,141 @@
+				};
+				return (
+					<VStack key={ purchase.ID }>
+						<HStack>
```
Should we align the element to the top?

arthur791004 commented on client/dashboard/me/billing-purchases/purchase-settings/upcoming-renewals-dialog.tsx:
```diff
@@ -0,0 +1,141 @@
+									) }
+								</Text>
+								{ showManagePurchaseLinks && (
+									<div className="upcoming-renewals-dialog__renewal-settings-link">
```
Would it be better to avoid it shrink? Also, it seems the class name is unused 🤔

sirbrillig replied:
Ah, thanks. I missed this when copying from the original component.

arthur791004 commented on client/dashboard/me/billing-purchases/purchase-settings/purchase-notice.tsx:
```diff
@@ -0,0 +1,363 @@
+				sprintf(
+					// translators: cardType is a credit card brand, cardNumber is the last 4 digits of the credit card number, and cardExpiry is the card expiration date.
+					__(
+						'Your %(cardType)s ending in %(cardNumber)d expires %(cardExpiry)s – before the next renewal. Please {{a}}update your payment information{{/a}}.'
```
Should it be `<a>...</a>` instead of `{{a}}...{{/a}}`

sirbrillig replied:
Oops. Thank you for spotting that!

arthur791004 commented on client/dashboard/me/billing-purchases/purchase-settings/purchase-notice.tsx:
```diff
@@ -0,0 +1,363 @@
+		: differenceInCalendarDays( new Date( purchase.expiry_date ), new Date() );
+	const productType =
+		purchase.product_slug === DotcomPlans.ECOMMERCE_TRIAL_MONTHLY
+			? __( 'ecommerce' )
```
I'm not sure but should it be `Ecommerce`?

arthur791004 commented on client/dashboard/me/billing-purchases/purchase-settings/purchase-expiring-notice.tsx:
```diff
@@ -0,0 +1,308 @@
+		return sprintf(
+			// translators: purchaseName is the name of the plan and expiry is a formatted string like "in 3 months".
+			__(
+				"%(purchaseName)s will expire and be removed from your site %(expiry)s. Please renew before expiry so you don't lose out on your paid features!"
```
We should use `‘` instead of `'`

arthur791004 commented on client/dashboard/me/billing-purchases/purchase-settings/purchase-expiring-notice.tsx:
```diff
@@ -0,0 +1,308 @@
+					{ purchaseName, expiry }
+				),
+				{
+					link: <Button onClick={ autoRenewingUpgradesAction } />,
```
Should it be `Link`?

sirbrillig replied:
I don't think so... clicking this text should display the upcoming renewals dialog box, which is definitely a button-like behavior (it doesn't link anywhere). That said, the button should be styled as a link, which is a good point. I'll update these.

sirbrillig committed:
Prevent RenewNoticeAction from returning null

sirbrillig committed:
Fix incorrect syntax for createInterpolateElement

sirbrillig committed:
Use smart quotes for don't

sirbrillig committed:
Update buttons in ExpiringLaterText to use link variant

sirbrillig committed:
Move price to its own column in UpcomingRenewalsDialog

sirbrillig committed:
Rename showManagePurchaseLinks prop to avoid default true

sirbrillig committed:
Close UpcomingRenewalsDialog when navigating away

sirbrillig committed:
Align UpcomingRenewalsDialog rows to top

sirbrillig committed:
Fix casing of Ecommerce plan

sirbrillig committed:
Replace list with DataView

sirbrillig committed:
Enable checkboxes in DataViews

sirbrillig committed:
Move "Manage purchase" to actions

Note: this is now suffering from this bug which prevents the actions
menu from being visible: https://github.com/wordpress/gutenberg/issues/71463

We need to find a way to reduce the z-index of ConfirmDialog

sirbrillig committed:
Lower the z-index of the ConfirmDialog in UpcomingRenewalsDialog

sirbrillig committed:
Use meta approach for snackbar notices

arthur791004 commented on client/dashboard/me/billing-purchases/purchase-settings/upcoming-renewals-dialog.tsx:
```diff
@@ -0,0 +1,203 @@
+
+	const fields = useMemo( () => getPurchaseFields(), [] );
+
+	const dataWithIds = useMemo(
```
I'm curious about whether it's required 🤔

sirbrillig replied:
Yeah, this was recommended by Claude but I think it's not actually needed.

arthur791004 commented on client/dashboard/me/billing-purchases/purchase-settings/other-renewable-purchases-notice.tsx:
```diff
@@ -0,0 +1,732 @@
+				<Notice variant={ noticeStatus }>
+					{ noticeText }
+					{ ( noticeActionHref || noticeActionOnClick ) && (
+						<Button
```
It should be the `actions` props of the `Notice`. Otherwise, it will be

<img width="679" height="110" alt="Image" src="https://github.com/user-attachments/assets/202ec33d-2e5a-42a6-8b7b-16aa0ac2dd23" />

arthur791004 commented on client/dashboard/me/billing-purchases/purchase-settings/other-renewable-purchases-notice.tsx:
```diff
@@ -0,0 +1,732 @@
+	) {
+		if ( ! currentPurchaseCreditCardExpiresBeforeSubscription ) {
+			noticeStatus = 'success';
+			noticeText = createInterpolateElement(
```
Would it be better to create a function to get the content and return early? It seems complex that we still need to continue to check everything when we already have the text, status, etc

sirbrillig replied:
Yes. The lack of early returns is very frustrating. I just migrated it in its current form to avoid having to test every scenario, but you're right. I'm going to refactor all of these.

sirbrillig committed:
Remove dataWithIds again

sirbrillig committed:
Move OtherRenewablePurchasesNotice button to Notice action prop

sirbrillig committed:
Refactor to use early returns for OtherRenewablePurchasesNotice

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/20325507

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @sirbrillig for including a screenshot in the description! This is really helpful for our translators.
